plugins/in_winevtlog/pack.c

```diff
@@ -0,0 +1,601 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include <fluent-bit/flb_compat.h>
+#include <fluent-bit/flb_mem.h>
+#include <fluent-bit/flb_utils.h>
+#include <fluent-bit/flb_pack.h>
+#include <fluent-bit/flb_input_plugin.h>
+#include <msgpack.h>
+#include <sddl.h>
+#include <locale.h>
+#include "winevtlog.h"
+
+#define FORMAT_ISO8601 "%Y-%m-%d %H:%M:%S %z"
+
+#define BINDATA(evt) ((unsigned char *) (evt) + (evt)->DataOffset)
+
+static void pack_nullstr(msgpack_packer *mp_pck)
+{
+    msgpack_pack_str(mp_pck, 0);
+    msgpack_pack_str_body(mp_pck, "", 0);
+}
+
+static int pack_wstr(msgpack_packer *mp_pck, const wchar_t *wstr, struct winevtlog_config *ctx)
+{
+    int size;
+    char *buf;
+    UINT code_page = CP_UTF8;
+    LPCSTR defaultChar = L" ";
+    if (ctx->use_ansi) {
+        code_page = CP_ACP;
+    }
+
+    /* Compute the buffer size first */
+    size = WideCharToMultiByte(code_page, 0, wstr, -1, NULL, 0, NULL, NULL);
+    if (size == 0) {
+        return -1;
+    }
+
+    buf = flb_malloc(size);
+    if (buf == NULL) {
+        flb_errno();
+        return -1;
+    }
+
+    /* Convert UTF-16 into UTF-8 */
+    size = WideCharToMultiByte(code_page, 0, wstr, -1, buf, size, defaultChar, NULL);
+    if (size == 0) {
+        flb_free(buf);
+        return -1;
+    }
+
+    /* Pack buf except the trailing '\0' */
+    msgpack_pack_str(mp_pck, size - 1);
+    msgpack_pack_str_body(mp_pck, buf, size - 1);
+    flb_free(buf);
+    return 0;
+}
+
+static int pack_binary(msgpack_packer *mp_pck, PBYTE bin, size_t length)
+{
+    const char *HEX_TABLE = "0123456789ABCDEF";
+    char *buffer;
+    int size = length * 2;
+    size_t i, j;
+    unsigned int idx = 0;
+
+    if (length == 0) {
+        pack_nullstr(mp_pck);
+        return 0;
+    }
+
+    buffer = flb_malloc(size);
+    if (buffer == NULL) {
+        flb_errno();
+        return -1;
+    }
+
+    for (i = 0; i < length; i++) {
+        for (j = 0; j < 2; j++) {
+            idx = (unsigned int)(bin[i] >> (j * 4) & 0x0F);
+            buffer[2*i+(1-j)] = HEX_TABLE[idx];
+        }
+    }
+    msgpack_pack_str(mp_pck, size);
+    msgpack_pack_str_body(mp_pck, buffer, size);
+    flb_free(buffer);
+
+    return 0;
+}
+
+static int pack_guid(msgpack_packer *mp_pck, const GUID *guid, struct winevtlog_config *ctx)
+{
+    LPOLESTR p = NULL;
+    if (FAILED(StringFromCLSID(guid, &p))) {
+        return -1;
+    }
+    if (pack_wstr(mp_pck, p, ctx)) {
+        CoTaskMemFree(p);
+        return -1;
+    }
+
+    CoTaskMemFree(p);
+
+    return 0;
+}
+
+static int pack_hex32(msgpack_packer *mp_pck, int32_t hex)
+{
+    CHAR buffer[32];
```

Comment:
Can we do something about spacing? as in adding it to separate conditional blocks, variable definition from code, etc?